pennylane/math/quantum.py

```diff
@@ -455,3 +456,123 @@ def to_density_matrix(state, indices, check_state=False, c_dtype="complex64"):
     density_matrix = _density_matrix_from_matrix(state, indices, check_state)
 
     return density_matrix
+
+
+def to_vn_entropy(state, indices=None, base=None, check_state=False, c_dtype="complex128"):
+    """Compute the Von Neumann entropy from a state vector, a density matrix given a subsystem.
+
+    Args:
+        state (tensor_like): ``(2**N)`` tensor state vector or ``(2**N, 2**N)`` tensor density matrix.
+        indices (list(int)): List of indices in the considered subsystem.
+        base (float, int): Base for the logarithm.
+        check_state (bool): If True, the function will check the state validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Von Neumann entropy of the considered subsystem.
+
+    **Example**
+
+    >>> x = [1, 0, 0, 1] / np.sqrt(2)
+    >>> to_vn_entropy(x, indices=[0])
+    0.6931472
+
+    >>> to_vn_entropy(x, indices=[0], base=2)
+    1.0
+
+    >>> y = [[1/2, 0, 0, 1/2], [0, 0, 0, 0], [0, 0, 0, 0], [1/2, 0, 0, 1/2]]
+    >>> to_vn_entropy(x, indices=[0])
+    0.6931472
+
+    """
+    state = cast(state, dtype=c_dtype)
+    len_state = state.shape[0]
+
+    if state.shape == (len_state,):
+        density_matrix = to_density_matrix(state, indices, check_state)
+        entropy = compute_vn_entropy(density_matrix, base)
+        return entropy
+
+    density_matrix = to_density_matrix(state, indices, check_state)
+    entropy = compute_vn_entropy(density_matrix, base)
+
+    return entropy

```

Comment:
This looks like the same code in both cases. It's probably best to simplify.

Reply:
I will simplify it 👍